### PR TITLE
List roles by resource_type

### LINKED ISSUE #288: Implement the ListRoles by resource_type endpoint
1.  listing only the roles for a specific resource type by setting the `resource_type` parameter. *done* as part of this task.

2. Listing all roles will be done in #348 

See [this section](https://docs.google.com/document/d/1qmpH8Yex0yKmcTauPqVoGmquRiEbab1E8fFf90tEw3A/edit#heading=h.kt11mbk2f6hw) of the detailed design document for details.

```
GET /api/role HTTP/1.1
Authorization: Bearer MHg3OUZEQkZBMjcx
```

```
HTTP/1.1 200 OK
Content-Type: application/json
...
[{  
   "role_id":"JA7X-PV9Z",
   "name":"collaborator",
   "resource_type":"openshift.io/resource/area"
   "scopes":["manage","view"]
},
{  
   "role_id":"P9A5-MX3H",
   "name":"administrator"
   "resource_type":"openshift.io/resource/area"
   "scopes":["admin"]
}]
```

### PR DESCRIPTION
**Pull request by sbose78**
fixes #288 

This PR implements the 
```
GET /api/roles?resource_type=NAME_OF_RESOURCE_TYPE

HTTP/1.1 200 OK
Content-Type: application/json
...
"data": [
  {
    "name": "collaborator",
    "resource_type": "openshift.io/resource/area",
    "scopes": [
      "manage",
      "view"
    ]
  },
  {
    "name": "administrator",
    "resource_type": "openshift.io/resource/area",
    "scopes": [
      "admin"
    ]
  }
]
```

Support for `GET /api/roles` will be added in a different PR.

- [x] Repository for Role/Scope associations in role_scope table
- [x] Repository tests
- [x] Model RoleManagementModel
- [x] Model tests
- [x] Service RoleManagementService
- [x] Service tests
- [x] Controller
- [x] Controller tests
- [x] Migrations / add indexes to columns
- [x] Migration tests

### CODE REVIEW

**Comment by fabric8cd** [comment]
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-347-1`

**Comment by fabric8cd** [comment]
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-347-2`

**Comment by fabric8cd** [comment]
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-347-3`

File: authorization/role/model/role_scope.go [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package model
+
+type RoleScope struct {
```
Comment (sbose78) [inline authorization/role/model/role_scope.go]: might remove this and end up using the `RoleScope` from `role/scope/repository/role_scope.go`

**Comment by fabric8cd** [comment]
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-347-5`

**Comment by fabric8cd** [comment]
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-347-6`

**Comment by fabric8cd** [comment]
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-347-7`

**Comment by fabric8cd** [comment]
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-347-9`

**Comment by fabric8cd** [comment]
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-347-10`

**Comment by fabric8cd** [comment]
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-347-12`

**Comment by fabric8cd** [comment]
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-347-13`

File: controller/roles_blackbox_test.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+type TestRolesRest struct {
+	gormtestsupport.DBTestSuite
+}
+
+func (s *TestRolesRest) SetupSuite() {
+	s.DBTestSuite.SetupSuite()
```
Comment (alexeykazakov) [inline controller/roles_blackbox_test.go]: No need to explicitly declare SetupSuite() if it only calls s.DBTestSuite.SetupSuite()

File: design/role.go [hunk 3]
```diff
@@ -5,6 +5,45 @@ import (
...
+			a.GET(""),
+		)
+		a.Params(func() {
+			a.Param("resource_type", d.String, "resource type for which roles are being listed")
+		})
+		a.Description("List available roles by resource type")
+		a.Response(d.OK, rolesMedia)
+		a.Response(d.InternalServerError, JSONAPIErrors)
+		a.Response(d.BadRequest, JSONAPIErrors)
+		a.Response(d.NotFound, JSONAPIErrors)
```
Comment (alexeykazakov) [inline design/role.go]: plus Unauthorized

File: design/role.go [hunk 4]
```diff
@@ -5,6 +5,45 @@ import (
...
+	})
+})
+
+var rolesMedia = a.MediaType("application/vnd.roles+json", func() {
+	a.Description("Available Roles for a Resource Type")
+	a.Attributes(func() {
+		// keeping one level of nesting so that it's easier to add metadata in future.
+		a.Attribute("data", a.ArrayOf(rolesData))
```
Comment (alexeykazakov) [inline design/role.go]: Fix the PR description. It returns data attribute. Not an array directly.

File: migration/sql-files/022-resource-type-index.sql [hunk 5]
```diff
@@ -0,0 +1,20 @@
+-- avoid the sequence scan 
+CREATE INDEX IF NOT EXISTS idx_name_rt_name ON resource_type(name);
+
+--- The ROLE_SCOPE table had a composite primary key (PK) of ( role_id,scope_id ) 
+--- As part of the following change, a new ID column is added which is made the PK.
```
Comment (alexeykazakov) [inline migration/sql-files/022-resource-type-index.sql]: "which is made the PK"  ?

File: migration/sql-files/022-resource-type-index.sql [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+--- The old PK ie ( role_id,scope_id ) is being made a Unique Key.
```
Comment (alexeykazakov) [inline migration/sql-files/022-resource-type-index.sql]: ie ?

File: migration/sql-files/022-resource-type-index.sql (hunk 5 above)
Comment (sbose78) [reply migration/sql-files/022-resource-type-index.sql]: the new ID column is made the primary key.

File: migration/sql-files/022-resource-type-index.sql (hunk 6 above)
Comment (sbose78) [reply migration/sql-files/022-resource-type-index.sql]: ie --> i.e. --> That is :D 
Sorry for the confusion.

File: migration/sql-files/022-resource-type-index.sql (hunk 5 above)
Comment (alexeykazakov) [reply migration/sql-files/022-resource-type-index.sql]: You mean the ID is now the PK, right?

File: authorization/role/model/role_management_model_service_backbox_test.go [hunk 7]
```diff
@@ -23,7 +35,11 @@ func TestRunroleManagementModelServiceBlackboxTest(t *testing.T) {
 
 func (s *roleManagementModelServiceBlackboxTest) SetupTest() {
 	s.DBTestSuite.SetupTest()
-	s.repo = rolemodel.NewRoleManagementModelService(s.DB, s.Application)
+	s.DB = s.DB.Debug()
```
[the remark this answers is not in the document]
Comment (sbose78) [reply authorization/role/model/role_management_model_service_backbox_test.go]: b550a4c
Yeah good idea, if it gets too noisy, we can consider turning it off later - but for now kept it in the `gormtestsupport.DBTestSuite`

File: controller/roles_blackbox_test.go (hunk 2 above)
Comment (sbose78) [reply controller/roles_blackbox_test.go]: b550a4c

File: design/role.go (hunk 3 above)
Comment (sbose78) [reply design/role.go]: 45f82fd12f13c96f1186470ee81d101578f91a22 thanks.

File: migration/sql-files/022-resource-type-index.sql (hunk 5 above)
Comment (sbose78) [reply migration/sql-files/022-resource-type-index.sql]: correct.

File: design/role.go (hunk 3 above)
Comment (alexeykazakov) [reply design/role.go]: Sure. It’s just for swagger docs.

**Comment by fabric8cd** [comment]
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-347-14`

File: authorization/role/service/role_management_service_blackbox_test.go [hunk 8]
```diff
@@ -238,3 +251,130 @@ func (s *roleManagementServiceBlackboxTest) compare(createdRole identityrole.Ide
...
+		require.Equal(s.T(), rt.ResourceTypeID, existingRole.ResourceTypeID)
+		for _, sc := range r.Scopes {
+			foundScope, err := s.checkScopeBelongsToResourceType(s.DB, sc, rt)
+			require.NoError(s.T(), err)
+			require.Equal(s.T(), true, foundScope)
+		}
+	}
+}
+
+func (s *roleManagementServiceBlackboxTest) checkScopeBelongsToResourceType(db *gorm.DB, scopeName string, rt resourcetype.ResourceType) (bool, error) {
```
Comment (sbryzak) [inline authorization/role/service/role_management_service_blackbox_test.go]: Querying all the resource type scopes just by name, and then iterating through them all to check if one is of a certain resource type seems terribly inefficient, why don't we simply have a Lookup() method in the repository that can return a ResourceTypeScope object for a specified resource type and scope name?

File: authorization/role/model/role_scope.go (hunk 1 above)
Comment (sbryzak) [reply authorization/role/model/role_scope.go]: Shouldn't this go in the authorization/role package? i.e. same as [this](https://github.com/fabric8-services/fabric8-auth/blob/master/authorization/organization/organization.go)?

File: authorization/role/scope/repository/role_scope.go [hunk 9]
```diff
@@ -0,0 +1,129 @@
+package repository
```
Comment (sbryzak) [inline authorization/role/scope/repository/role_scope.go]: Also, I thought we decided to flatten our structure during last week's call, so shouldn't this just be in the authorization/role/repository directory?

File: authorization/role/service/role_management_service_blackbox_test.go [hunk 10]
```diff
@@ -238,3 +251,130 @@ func (s *roleManagementServiceBlackboxTest) compare(createdRole identityrole.Ide
 
 	return true
 }
+
+func (s *roleManagementServiceBlackboxTest) TestGetRolesByResourceTypeOKEmpty() {
```
Comment (sbryzak) [inline authorization/role/service/role_management_service_blackbox_test.go]: Really nice set of comprehensive tests for this :)

File: migration/sql-files/022-resource-type-index.sql [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+--- 2. The gormsupport/cleaner/db_clean.go depends on using the entity's KEY attribute to 
+--- clean up the rows after tests. When there's no single attribute that's a primary key, the implicit cleanup logic
+--- breaks. That ends up with tests producing a lot of data which don't get cleaned up in the end. This causes a cascade of failed
+--- cleanups, upto the RESOURCE_TYPE table.
+---
+--- Adding a new ID column helped avoid doing unnecesary refactoring of the cleanup code ( which worked for all other tables )
+--- without having to change the way we queried the ROLE_SCOPE table in the business/model/service code.
+
+ALTER TABLE ROLE_SCOPE DROP CONSTRAINT ROLE_SCOPE_PKEY;
+ALTER TABLE ROLE_SCOPE ADD ID UUID PRIMARY KEY DEFAULT uuid_generate_v4();
```
Comment (sbryzak) [inline migration/sql-files/022-resource-type-index.sql]: Replacing a natural primary key with a surrogate key is bad database design.  Is there any reason why cleanup can't happen with a composite key?

File: authorization/role/scope/repository/role_scope.go (hunk 9 above)
Comment (alexeykazakov) [reply authorization/role/scope/repository/role_scope.go]: I remember that we agreed on this too.

Comment (sbose78) [reply authorization/role/scope/repository/role_scope.go]: yeah, had missed doing so fccf380bc0733dae57d5a6f15145cd638cfa77df

**Comment by fabric8cd** [comment]
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-347-17`

**Comment by fabric8cd** [comment]
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-347-18`

File: migration/sql-files/022-resource-type-index.sql (hunk 11 above)
Comment (alexeykazakov) [reply migration/sql-files/022-resource-type-index.sql]: Here is a PR to support multiple PKs in cleanup - https://github.com/fabric8-services/fabric8-auth/pull/367

File: migration/migration_blackbox_test.go [hunk 12]
```diff
@@ -108,6 +108,7 @@ func TestMigrations(t *testing.T) {
 	t.Run("TestMigration18", testMigration18)
 	t.Run("TestMigration21", testMigration21)
 	t.Run("TestMigration22", testMigration22)
+	t.Run("TestMigration22", testMigration23)
```
[the remark this answers is not in the document]
Comment (sbose78) [reply migration/migration_blackbox_test.go]: thanks b6bfd55979e60611ea09e51faab342ace351dc6c

**Comment by sbose78** [comment]
@sbryzak  , reverted the changes in role_scope model after @alexeykazakov's fix.
87d9b0e76bcf1472b77b7a77e6b3f1755d36156d

**Comment by fabric8cd** [comment]
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-347-19`

**Comment by sbose78** [comment]
@sbryzak could you please provide your review on this PR ? would like to merge this soon :)

**Comment by fabric8cd** [comment]
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-347-20`

File: authorization/resourcetype/scope/repository/resource_type_scope.go [hunk 13]
```diff
@@ -196,3 +199,44 @@ func (m *GormResourceTypeScopeRepository) List(ctx context.Context, resourceType
...
+
+/*
+
+// ListByName returns all resource type scopes filtered by name
+func (m *GormResourceTypeScopeRepository) ListByName(ctx context.Context, name string) ([]ResourceTypeScope, error) {
+	return m.Query(FilterByScopeName(name))
+}
+*/
+// ListByResourceTypeAndScope returns all resource type scopes filtered by name and scope.
+func (m *GormResourceTypeScopeRepository) ListByResourceTypeAndScope(ctx context.Context, resourceTypeID uuid.UUID, name string) ([]ResourceTypeScope, error) {
```
Comment (sbryzak) [inline authorization/resourcetype/scope/repository/resource_type_scope.go]: What is the use case for this function?

Comment (sbose78) [reply authorization/resourcetype/scope/repository/resource_type_scope.go]: Can't seem to remember how this got here :) removing..
nice catch!

Comment (sbose78) [reply authorization/resourcetype/scope/repository/resource_type_scope.go]: they are actually used in the role model test https://github.com/fabric8-services/fabric8-auth/pull/347/files#diff-c033b4ea6b843569115b9d21a5249f39R169 , so I'll have to keep this function.

Comment (sbose78) [reply authorization/resourcetype/scope/repository/resource_type_scope.go]: It was added after your review comment https://github.com/fabric8-services/fabric8-auth/pull/347#discussion_r172490229

**Comment by fabric8cd** [comment]
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-347-21`

File: authorization/resourcetype/scope/repository/resource_type_scope.go (hunk 13 above)
Comment (sbryzak) [reply authorization/resourcetype/scope/repository/resource_type_scope.go]: Is it used outside of the tests though? I'm wondering what the purpose of this method is in production - to me, I think a Lookup() method that accepted the resource type and scope as parameters (so same as you have now), however returns the single scope record found, would be more useful.  How can there be multiple scopes returned if you're already providing both the resource type and the scope name?

Comment (sbose78) [reply authorization/resourcetype/scope/repository/resource_type_scope.go]: > I'm wondering what the purpose of this method is in production

It's not being exposed as a service, rather as a clean API - a wrapper over calling the raw sql.

> however returns the single scope record found, would be more useful

Good catch, let me fix this.

**Comment by fabric8cd** [comment]
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-347-22`

File: authorization/resourcetype/scope/repository/resource_type_scope.go (hunk 13 above)
Comment (sbose78) [reply authorization/resourcetype/scope/repository/resource_type_scope.go]: https://github.com/fabric8-services/fabric8-auth/pull/347/commits/46fe4512e9a3a57b025f2f7a4e963d3d2205c0c5

File: authorization/resourcetype/scope/repository/resource_type_scope.go [hunk 14]
```diff
@@ -196,3 +198,41 @@ func (m *GormResourceTypeScopeRepository) List(ctx context.Context, resourceType
 	}
 	return rows, nil
 }
+
+// ListByResourceTypeAndScope returns all resource type scopes filtered by name and scope.
+func (m *GormResourceTypeScopeRepository) ListByResourceTypeAndScope(ctx context.Context, resourceTypeID uuid.UUID, name string) (*ResourceTypeScope, error) {
```
Comment (sbryzak) [inline authorization/resourcetype/scope/repository/resource_type_scope.go]: List... suggests multiple results, I would rename this LookupByResourceTypeAndScope instead

Comment (sbose78) [reply authorization/resourcetype/scope/repository/resource_type_scope.go]: Done, @sbryzak . https://github.com/fabric8-services/fabric8-auth/pull/347/commits/fc766284c4cab87d941374432e47ea93e7eb3221  Thanks for pointing out !

**Comment by fabric8cd** [comment]
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-347-23`

**Comment by fabric8cd** [comment]
@sbose78 snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-347-24`

**Pull request merged**